parallax/probe_calibration.py

```diff
@@ -3,125 +3,296 @@
 
-        logger.debug(f"Error (Euclidean distance): {error:.5f}, min_error: {self.error_min:.5f}")
-        logger.debug(f"Transformed point: {transformed_point}")
-        logger.debug(f"Expected global point: {global_point}")
-        logger.debug(f"local points: {local_point}")
+        Returns:
+            bool: True if the error is within threshold, otherwise False.
+        """
+        if self.LR_err_L2_current <= self.LR_err_L2_threshold:
+            return True
+        else:
+            return False
 
-    def local_global_transform(self, stage):
-        """Perform local to global transformation.
-        
-        Args:
-            stage (Stage): Stage object containing stage coordinates.
+    def _enough_points_emit_signal(self):
         """
-        self.update(stage)
-        if self.is_enough_points():
-            local_points, global_points = self.reshape_array()
-            retval, transform_matrix, self.inliers = cv2.estimateAffine3D(local_points, global_points, \
-                                                            ransacThreshold = 30, confidence=0.995)
+        Emits calibration complete signals based on the sufficiency of point ranges in each direction.
+        """
+        if not self.signal_emitted_x and self.max_x - self.min_x > self.threshold_min_mix:
+            self.calib_complete_x.emit()
+            self.signal_emitted_x = True
+        if not self.signal_emitted_y and self.max_y - self.min_y > self.threshold_min_mix:
+            self.calib_complete_y.emit()
+            self.signal_emitted_y = True
+        if not self.signal_emitted_z and self.max_z - self.min_z > self.threshold_min_mix:
+            self.calib_complete_z.emit()
+            self.signal_emitted_z = True
 
-        if retval and transform_matrix is not None:
-            self._test_cmp_truth_expect(stage, transform_matrix)
-            logger.debug("========================")
-            local_point = np.array([10346.5, 14720.0, 8270.5, 1])
-            global_point = np.array([0.0, 0.0, 0.0])
-            transformed_point = np.dot(transform_matrix, local_point)[:3]
-            error = np.linalg.norm(transformed_point - global_point)
-            logger.debug(f"test. error: {error}")
-            logger.debug(f"Transformed point: {transformed_point}")
-            logger.debug(self.transform_matrix)
+    def _is_enough_points(self):
+        """Check if there are enough points for calibration.
+        
+        Returns:
+            bool: True if there are enough points, False otherwise.
+        """ 
+        # End Criteria: 
+        # 1. distance maxX-minX, maxY-minY, maxZ-minZ
+        # 2. transM_LR difference in some epsilon value
+        # 3. L2 error (Global and Exp) is less than some values (e.g. 20 mincrons)
+        if self._is_criteria_met_points_min_max():
+            logger.debug("_is_criteria_met_points_min_max True")
```

Comment:
Should there be a check so this only gets printed once?

Reply:
Fixed. Debug messages can be seen on the log file only by the programmer if it is set to 'logger.setLevel(logging.DEBUG)'. User cannot see those message from the terminal prompt. No need to worry too much for the debug msg. I changed the debug message anyway! 
===============================
        if self._is_criteria_met_points_min_max():
            if self._is_criteria_met_transM():
                if self._is_criteria_met_l2_error():
                    logger.debug("Enough points gathered.")
                    return True
===============================